# Name: Yu-Chen, Hsiao

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing  import StandardScaler
from sklearn.naive_bayes import MultinomialNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression
from sklearn import preprocessing
from sklearn.preprocessing import Normalizer
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import GaussianNB

In [2]:
df = pd.read_csv('C:/Users/Spyder/Documents/MobaXterm/home/Titanic/Data/jhetrick.io/ANLT210/titanic.csv')

In [3]:
sex_mapping = {
           'male': 1,
           'female': 0}
df['Sex'] = df['Sex'].map(sex_mapping)

In [4]:
df1 = df.drop('Name',axis = 1)

# Normalize

In [5]:
x = df1.values
min_max_scaler = preprocessing.MinMaxScaler()
x_scaled = min_max_scaler.fit_transform(x)
df1 = pd.DataFrame(x_scaled)

In [6]:
x = df1.iloc[:,1:]
y = df1.iloc[:,0]

# Logistic Regression

In [7]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=55)
lr=LogisticRegression().fit(x_train,y_train)
lr.predict(x_test)
sum(lr.predict(x_test) == y_test)/len(x_test)

0.8239700374531835

# Gaussian Navie Bayes

In [8]:
nb = GaussianNB().fit(x_train, y_train)
nb.predict(x_test)
sum(nb.predict(x_test) == y_test)/len(x_test)

0.8277153558052435

# K-nearest Neighbor

In [9]:
k = []
accuracy = []
for i in range(1,len(x_train)):
    knn = KNeighborsClassifier(n_neighbors=i).fit(x_train, y_train)
    knn.predict(x_test)
    acc = sum(knn.predict(x_test) == y_test)/len(x_test)
    k.append(i)
    accuracy.append(acc)

In [12]:
K = pd.DataFrame(k)
K['Accuracy'] = accuracy
K = K.rename(columns={0:'K'})
K.sort_values(by = 'Accuracy', ascending = False)
max(K['Accuracy'])

0.8539325842696629

# Linear Regression

In [13]:
reg = LinearRegression().fit(x, y)
sur = reg.predict(x_test)
s = []
for s1 in sur:
    if s1 < 0.5:
        s.append(0)
    else:
        s.append(1)
sum(s == y_test)/len(x_test)

0.8239700374531835

# Summary

I tried 4 models to predict the probability of surviving. The model with highest accuracy is Knn model while k equals to 15. The predicting accuracy is round 85%.